File: scripts/fov/FieldOfView.cs
using System;
using System.Collections.Generic;
using Godot;
using ProiectDeLegume.Scripts.World;

namespace ProiectDeLegume.Scripts.Fov;

/// <summary>
/// Fog of War with DDA raycasting. Port from KorGE version.
/// Renders as a Sprite2D with a per-tile Image overlay.
/// </summary>
public partial class FieldOfView : Sprite2D
{
    private const double FovAngle = 120.0;     // degrees
    private const int FovRadius = 12;          // tiles
    private const double RayStep = 0.5;        // degrees between rays
    private const int BehindRadius = 3;        // peripheral vision

    private const int TileSize = TileSetGenerator.TileSize;

    private int _mapW;
    private int _mapH;

    // 0 = unknown, 1 = explored, 2 = visible
    private int[] _visibility;

    private Image _fogImage;
    private ImageTexture _fogTexture;

    // Tile data for vision blocking
    private TileMapLayer _buildingLayer;
    private TileMapLayer _objectLayer;

    // Vision-blocking tile indices
    private static readonly HashSet<int> VisionBlockers = new()
    {
        TileSetGenerator.Wall,
        TileSetGenerator.Tree,
    };

    public void Init(int mapW, int mapH, TileMapLayer buildingLayer, TileMapLayer objectLayer)
    {
        _mapW = mapW;
        _mapH = mapH;
        _buildingLayer = buildingLayer;
        _objectLayer = objectLayer;

        _visibility = new int[_mapW * _mapH];

        _fogImage = Image.CreateEmpty(_mapW, _mapH, false, Image.Format.Rgba8);
        // Fill black
        for (int y = 0; y < _mapH; y++)
            for (int x = 0; x < _mapW; x++)
                _fogImage.SetPixel(x, y, new Color(0, 0, 0, 0.96f));

        _fogTexture = ImageTexture.CreateFromImage(_fogImage);
        Texture = _fogTexture;
        Centered = false; // origin = top-left, not center
        TextureFilter = TextureFilterEnum.Nearest;

        // Scale to cover the whole map (1 pixel per tile → TileSize pixels per tile)
        Scale = new Vector2(TileSize, TileSize);
        Position = Vector2.Zero;
        ZIndex = 5; // above tiles, below player
    }

    public void UpdateFov(double playerX, double playerY, double facingAngle, double fovMultiplier = 1.0)
    {
        int playerTileX = (int)(playerX / TileSize);
        int playerTileY = (int)(playerY / TileSize);
        int effectiveRadius = Math.Max(3, (int)(FovRadius * fovMultiplier));

        // Demote visible -> explored
        for (int i = 0; i < _mapW * _mapH; i++)
        {
            if (_visibility[i] == 2) _visibility[i] = 1;
        }

        // Peripheral vision (short rays in all directions, wall-aware)
        for (int deg = 0; deg < 360; deg += 2)
        {
            CastRay(playerX, playerY, deg * Math.PI / 180.0, BehindRadius);
        }

        // FOV cone rays
        double halfFov = FovAngle / 2.0;
        double startAngle = facingAngle * 180.0 / Math.PI - halfFov;
        double endAngle = facingAngle * 180.0 / Math.PI + halfFov;

        for (double angle = startAngle; angle <= endAngle; angle += RayStep)
        {
            CastRay(playerX, playerY, angle * Math.PI / 180.0, effectiveRadius);
        }

        // Update fog image
        UpdateFogImage();
    }

    private void CastRay(double originX, double originY, double angleRad, int maxTiles)
    {
        double dirX = Math.Cos(angleRad);
        double dirY = Math.Sin(angleRad);
        double ts = TileSize;
        double maxDistSq = (maxTiles * ts) * (maxTiles * ts);

        int tileX = (int)(originX / ts);
        int tileY = (int)(originY / ts);

        int stepX = dirX >= 0 ? 1 : -1;
        int stepY = dirY >= 0 ? 1 : -1;

        double tDeltaX = dirX != 0 ? Math.Abs(ts / dirX) : double.MaxValue;
        double tDeltaY = dirY != 0 ? Math.Abs(ts / dirY) : double.MaxValue;

        double tMaxX = dirX != 0
            ? Math.Abs(((dirX > 0 ? (tileX + 1) * ts : tileX * ts) - originX) / dirX)
            : double.MaxValue;
        double tMaxY = dirY != 0
            ? Math.Abs(((dirY > 0 ? (tileY + 1) * ts : tileY * ts) - originY) / dirY)
            : double.MaxValue;

        int steps = 0;
        int maxSteps = maxTiles * 3;

        while (steps < maxSteps)
        {
            if (tileX < 0 || tileX >= _mapW || tileY < 0 || tileY >= _mapH) break;

            double cx = tileX * ts + ts / 2 - originX;
            double cy = tileY * ts + ts / 2 - originY;
            if (cx * cx + cy * cy > maxDistSq) break;

            _visibility[tileY * _mapW + tileX] = 2;

            // Stop at vision-blocking tiles
            if (BlocksVision(tileX, tileY)) break;

            if (tMaxX < tMaxY)
            {
                tMaxX += tDeltaX;
                tileX += stepX;
            }
            else
            {
                tMaxY += tDeltaY;
                tileY += stepY;
            }
            steps++;
        }
    }

    private bool BlocksVision(int x, int y)
    {
        var pos = new Vector2I(x, y);

        // Check building layer
        var buildingCoords = _buildingLayer.GetCellAtlasCoords(pos);
        if (buildingCoords.X >= 0 && VisionBlockers.Contains(buildingCoords.X))
            return true;

        // Check object layer
        var objectCoords = _objectLayer.GetCellAtlasCoords(pos);
        if (objectCoords.X >= 0 && VisionBlockers.Contains(objectCoords.X))
            return true;

        return false;
    }

    private void UpdateFogImage()
    {
        for (int y = 0; y < _mapH; y++)
        {
            for (int x = 0; x < _mapW; x++)
            {
                int v = _visibility[y * _mapW + x];
                Color color = v switch
                {
                    2 => new Color(0, 0, 0, 0),        // visible — clear
                    1 => new Color(0, 0, 0, 0.55f),    // explored — dimmed
                    _ => new Color(0, 0, 0, 0.96f),    // unknown — nearly black
                };
                _fogImage.SetPixel(x, y, color);
            }
        }

        _fogTexture.Update(_fogImage);
    }
}

File: src/commonMain/kotlin/fov/FieldOfView.kt
package fov

import korlibs.image.bitmap.*
import korlibs.image.color.*
import korlibs.korge.view.*
import korlibs.math.geom.*
import world.GameWorld
import kotlin.math.*

class FieldOfView(
    private val world: GameWorld,
    private val container: Container
) {
    companion object {
        const val FOV_ANGLE = 120.0          // degrees
        const val FOV_RADIUS = 12            // tiles
        const val RAY_STEP = 0.5             // degrees between rays
        const val BEHIND_RADIUS = 3          // small radius behind player
    }

    private val mapW = GameWorld.MAP_WIDTH
    private val mapH = GameWorld.MAP_HEIGHT
    private val tileSize = GameWorld.TILE_SIZE

    // Visibility states per tile
    // 0 = unknown (black), 1 = explored (dark grey), 2 = visible (clear)
    private val visibility = IntArray(mapW * mapH) { 0 }

    // Bitmap-based fog overlay: one pixel per tile, scaled up
    private val fogBitmap = Bitmap32(mapW, mapH, premultiplied = false)
    private val fogImage: Image

    init {
        // Fill fog completely black (opaque)
        for (i in 0 until mapW * mapH) {
            fogBitmap.setRgba(i % mapW, i / mapW, RGBA(0, 0, 0, 255))
        }
        fogImage = container.image(fogBitmap) {
            smoothing = false
            size(Size((mapW * tileSize).toFloat(), (mapH * tileSize).toFloat()))
        }
    }

    fun update(playerX: Double, playerY: Double, facingAngle: Double, fovMultiplier: Double = 1.0) {
        val effectiveRadius = (FOV_RADIUS * fovMultiplier).toInt().coerceAtLeast(3)
        val playerTileX = (playerX / tileSize).toInt()
        val playerTileY = (playerY / tileSize).toInt()

        // Demote all currently visible tiles to explored
        for (i in 0 until mapW * mapH) {
            if (visibility[i] == 2) {
                visibility[i] = 1
            }
        }

        // Small circle around player — wall-aware peripheral vision
        // Cast short rays in all directions to fill the close area
        for (deg in 0 until 360 step 2) {
            castRay(playerX, playerY, Math.toRadians(deg.toDouble()), maxTiles = BEHIND_RADIUS)
        }

        // Cast rays within the FOV cone
        val halfFov = FOV_ANGLE / 2.0
        val startAngle = Math.toDegrees(facingAngle) - halfFov
        val endAngle = Math.toDegrees(facingAngle) + halfFov

        var angle = startAngle
        while (angle <= endAngle) {
            castRay(playerX, playerY, Math.toRadians(angle), maxTiles = effectiveRadius)
            angle += RAY_STEP
        }

        // Update fog bitmap
        updateFogBitmap()
    }

    /**
     * DDA raycasting — visits every tile the ray passes through,
     * so it never skips over a 1-tile wall.
     */
    private fun castRay(originX: Double, originY: Double, angleRad: Double, maxTiles: Int = FOV_RADIUS) {
        val dirX = cos(angleRad)
        val dirY = sin(angleRad)
        val ts = tileSize.toDouble()
        val maxDistSq = (maxTiles * ts) * (maxTiles * ts)

        // Current tile
        var tileX = (originX / ts).toInt()
        var tileY = (originY / ts).toInt()

        // Step direction (+1 or -1)
        val stepX = if (dirX >= 0) 1 else -1
        val stepY = if (dirY >= 0) 1 else -1

        // Distance along ray to next vertical/horizontal tile boundary
        val tDeltaX = if (dirX != 0.0) abs(ts / dirX) else Double.MAX_VALUE
        val tDeltaY = if (dirY != 0.0) abs(ts / dirY) else Double.MAX_VALUE

        // Initial distance to first boundary
        var tMaxX = if (dirX != 0.0) {
            val boundary = if (dirX > 0) (tileX + 1) * ts else tileX * ts
            abs((boundary - originX) / dirX)
        } else Double.MAX_VALUE

        var tMaxY = if (dirY != 0.0) {
            val boundary = if (dirY > 0) (tileY + 1) * ts else tileY * ts
            abs((boundary - originY) / dirY)
        } else Double.MAX_VALUE

        var steps = 0
        val maxSteps = maxTiles * 3 // safety limit

        while (steps < maxSteps) {
            if (tileX < 0 || tileX >= mapW || tileY < 0 || tileY >= mapH) break

            // Check distance from origin
            val cx = tileX * ts + ts / 2 - originX
            val cy = tileY * ts + ts / 2 - originY
            if (cx * cx + cy * cy > maxDistSq) break

            visibility[tileY * mapW + tileX] = 2

            // Stop at vision-blocking tiles (walls, trees) but mark them visible
            if (world.blocksVision(tileX, tileY)) break

            // Step to next tile
            if (tMaxX < tMaxY) {
                tMaxX += tDeltaX
                tileX += stepX
            } else {
                tMaxY += tDeltaY
                tileY += stepY
            }
            steps++
        }
    }

    private fun updateFogBitmap() {
        fogBitmap.lock()
        for (y in 0 until mapH) {
            for (x in 0 until mapW) {
                val v = visibility[y * mapW + x]
                val color = when (v) {
                    2 -> RGBA(0, 0, 0, 0)         // visible — fully clear
                    1 -> RGBA(0, 0, 0, 140)       // explored — dimmed but visible
                    else -> RGBA(0, 0, 0, 245)    // unknown — almost black
                }
                fogBitmap.setRgba(x, y, color)
            }
        }
        fogBitmap.unlock()
        fogImage.bitmap = fogBitmap.slice()
    }
}
